### PR TITLE
[FLINK-24489][CEP] The size of entryCache & eventsBufferCache in the SharedBuffer should be defined with a threshold to limit the number of the elements in the cache

### PR DESCRIPTION
**Pull request by RocMarshal**
## What is the purpose of the change

The size of entryCache & eventsBufferCache in the SharedBuffer should be defined with a threshold to limit the number of the elements in the cache

## Brief change log

- Introduce `CEPCacheOptions` class.
- Replace the HashMap with google guava LocalCache in `SharedBuffer`. 
- The number of elements in the new cache will be limited by the configuration items from job global-parameters

## Verifying this change

- add ITest cases in `CEPITCase`
- add Test cases in `SharedBufferTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by RocMarshal** [comment]
@Airblader Could you help me to review this pr ? Thank you.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/configuration/CEPCacheOptions.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.time.Duration;
+
+/** CEP Cache Options. */
+public class CEPCacheOptions {
+
+    private CEPCacheOptions() {}
+
+    public static final ConfigOption<Integer> CEP_SHARED_BUFFER_EVENT_CACHE_SLOTS =
+            ConfigOptions.key("pipeline.global-job-parameters.cep.sharedbuffer.event-cache-slots")
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/configuration/CEPCacheOptions.java:33]: The cache options seem to be described way too technical. How should an end-user understand what's going on?

**Review by AHeise — CHANGES_REQUESTED** [review]
What are you trying to solve? Could you please fill out the PR form correctly? I understand what your are doing on technical side but I don't understand why it's necessary. The ticket is also not giving any context information whatsoever.

I'm also interested in the side-effects. If we remove events prematurely do we miss certain patterns? Are we trading memory for correctness here?

You also should supply a test for your change.

**Comment by RocMarshal** [comment]
> What are you trying to solve? Could you please fill out the PR form correctly? I understand what your are doing on technical side but I don't understand why it's necessary. The ticket is also not giving any context information whatsoever.
> 
> I'm also interested in the side-effects. If we remove events prematurely do we miss certain patterns? Are we trading memory for correctness here?
> 
> You also should supply a test for your change.

@AHeise I'm very sorry for such hasty description for the jira ticket and the PR. I uploaded the jvm memory dump file analyze result in the ticket and added a comment. I'm much obliged to you for helping me looking into the  significance of the ticket.
And after the discussion result of the ticket , I'll make response in closing it or re-filling out the PR form to advance it for the PR.
Thanks again.

**Comment by AHeise** [comment]
The situation is a bit odd. I think we shouldn't have a cache on top of rocks DB at all. We could think about having a cache in rocks DB itself.

Nevertheless, the cache is there and causes trouble for you. I understood now that your changes are not impacting correctness but only performance, so I think your contribution certainly is welcome.

To reiterate from my first rough review: Could you please add a (unit) test and make the configuration descriptions easier to understand for folks that have not read the CEP code?

**Comment by RocMarshal** [comment]
> The situation is a bit odd. I think we shouldn't have a cache on top of rocks DB at all. We could think about having a cache in rocks DB itself.
> 
> Nevertheless, the cache is there and causes trouble for you. I understood now that your changes are not impacting correctness but only performance, so I think your contribution certainly is welcome.
> 
> To reiterate from my first rough review: Could you please add a (unit) test and make the configuration descriptions easier to understand for folks that have not read the CEP code?

Thanks so much for your @AHeise  reply and explanation, which make me suddenly see the light.
As updated in the `CEPCacheOptions`: 
`
And it could accelerate the CEP operate process  speed and limit the capacity of cache in pure memory. Note: It's only effective to  limit usage of memory when 'state.backend' was set as 'rocksdb', which would  transport the elements exceeded the number of the cache into the rocksdb state storage instead of memory state storage.
`
The minor pr is really helpful for memory limitation when the `state.backend` is set as `rocksdb`.
By contrast，when the `state.backend` is set as not `rocksdb`, the current cache in the pr would cause performance decreased. Compared with old cache , the state part will contain more elements swapped out from new guava-cache ,which would make it  heavier to `copy on write`  for state .

In my limited read, if the positive significance of the pr is bigger than the side-effects of the pr, it would be worthy to do.
And Please let me know what do you think of it.

**Comment by RocMarshal** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
> In my limited read, if the positive significance of the pr is bigger than the side-effects of the pr, it would be worthy to do. And Please let me know what do you think of it.

Yes, I agree. So if you could address my initial comments (making the description more easy to understand for high-level CEP users), then I could merge it.

**Comment by RocMarshal** [comment]
@AHeise Thank you very much for your help.

**Review by AHeise — COMMENTED** [review]
Oops, it seems that my latest comment was not published :/ Sorry for the delay.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 2]
```diff
@@ -73,33 +88,105 @@
...
 
+    @VisibleForTesting
     public SharedBuffer(KeyedStateStore stateStore, TypeSerializer<V> valueSerializer) {
+        this(stateStore, valueSerializer, new Configuration());
+    }
+
+    public SharedBuffer(
+            KeyedStateStore stateStore,
+            TypeSerializer<V> valueSerializer,
+            ExecutionConfig.GlobalJobParameters globalJobParameters) {
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: @AHeise Thanks a lot for the reply. And I updated it bases on your comments. PTAL.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java [hunk 3]
```diff
@@ -18,31 +18,47 @@
...
+import static org.apache.flink.cep.utils.TestSharedBuffer.MINI_CACHE_CONFIG;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
 /** Tests for {@link SharedBuffer}. */
+@RunWith(Parameterized.class)
 public class SharedBufferTest extends TestLogger {
 
+    @Parameterized.Parameter public ExecutionConfig.GlobalJobParameters jobParameters;
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java:51]: Can we directly use `SharedBufferCacheConfig` here?

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/utils/TestSharedBuffer.java [hunk 4]
```diff
@@ -30,27 +31,57 @@
...
+                CEPCacheOptions.CEP_CACHE_STATISTICS_INTERVAL,
+                Duration.ofMillis(TimeUnit.SECONDS.toMillis(1)));
+
+        BIG_CACHE_CONFIG.set(CEPCacheOptions.CEP_SHARED_BUFFER_ENTRY_CACHE_SLOTS, 10240);
+        BIG_CACHE_CONFIG.set(CEPCacheOptions.CEP_SHARED_BUFFER_EVENT_CACHE_SLOTS, 10240);
+        BIG_CACHE_CONFIG.set(
+                CEPCacheOptions.CEP_CACHE_STATISTICS_INTERVAL,
+                Duration.ofMillis(TimeUnit.SECONDS.toMillis(1)));
+    }
+
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/utils/TestSharedBuffer.java:69]: Use `SharedBufferCacheConfig`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/configuration/SharedBufferCacheConfig.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            final int eventsBufferCacheSlots,
+            final int entryCacheSlots,
+            final Duration cacheStatisticsInterval) {
+        this.cacheStatisticsInterval = cacheStatisticsInterval;
+        this.entryCacheSlots = entryCacheSlots;
+        this.eventsBufferCacheSlots = eventsBufferCacheSlots;
+    }
+
+    public static SharedBufferCacheConfig of(
+            ExecutionConfig.GlobalJobParameters globalJobParameters) {
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/configuration/SharedBufferCacheConfig.java:65]: Please use `ReadableConfig` here. `GlobalJobParameters` is only for pure user code.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/CepOperator.java [hunk 6]
```diff
@@ -180,7 +181,12 @@ public void initializeState(StateInitializationContext context) throws Exception
                                 new ValueStateDescriptor<>(
                                         NFA_STATE_NAME, new NFAStateSerializer()));
 
-        partialMatches = new SharedBuffer<>(context.getKeyedStateStore(), inputSerializer);
+        partialMatches =
+                new SharedBuffer<>(
+                        context.getKeyedStateStore(),
+                        inputSerializer,
+                        SharedBufferCacheConfig.of(
+                                cepRuntimeContext.getExecutionConfig().getGlobalJobParameters()));
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/CepOperator.java:189]: `getOperatorConfig().getConfiguration()`

**Review by AHeise — CHANGES_REQUESTED** [review]
Looks much better already. Now let's get rid of `GlobalJobParameters` entirely. That is a concept only meant for user code and can't be used together with `flink-conf.yaml`.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java [hunk 7]
```diff
@@ -37,12 +43,25 @@
...
 /** Tests for {@link SharedBuffer}. */
+@RunWith(Parameterized.class)
 public class SharedBufferTest extends TestLogger {
 
+    public static final SharedBufferCacheConfig MINI_CACHE_CONFIG =
+            new SharedBufferCacheConfig(1, 1, Duration.ofSeconds(1L));
+    public static final SharedBufferCacheConfig BIG_CACHE_CONFIG =
+            new SharedBufferCacheConfig(10240, 10240, Duration.ofSeconds(1L));
+
+    @Parameterized.Parameter public SharedBufferCacheConfig jobParameters;
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java:54]: nit: rename to cache config.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java [hunk 8]
```diff
@@ -289,7 +308,7 @@ public void testSharedBufferCountersClearing() throws Exception {
     @Test
     public void testSharedBufferAccessor() throws Exception {
         TestSharedBuffer<Event> sharedBuffer =
-                TestSharedBuffer.createTestBuffer(Event.createTypeSerializer());
+                TestSharedBuffer.createTestBuffer(Event.createTypeSerializer(), BIG_CACHE_CONFIG);
```
Comment (AHeise) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferTest.java:311]: This should probably use `jobParameters`.

**Review by AHeise — COMMENTED** [review]
For me this looks almost ready to merge. I found 2 more nits in the code. Then please also
- run `mvn -Pgenerate-config-docs install` to refresh the docs and get a green build
- squash all commits into a single commit

Thank you very much!

**Comment by AHeise** [comment]
Please rebase and squash, so that I can merge.

**Comment by RocMarshal** [comment]
@AHeise Thank you very much for your comments, which makes us closer to higher quality step by step. It's really great. I have benefited a lot.
I have been striving for the best improvement according to your suggestions, which makes me unable to reply to you in time. So sorry for this.
I added some documentation about the new configuration parameters, Maybe it could meet our will for the standard.

**Review by AHeise — APPROVED** [review]
LGTM. Thank you for your contribution!

**Pull request merged**